web/src/message_events.js

```diff
@@ -366,15 +350,44 @@ export function update_messages(events) {
                     moved_message.stream_id = new_stream_id;
                     moved_message.display_recipient = new_stream_name;
                 }
-
-                // Add the Recent Conversations entry for the new stream/topics.
-                stream_topic_history.add_message({
-                    stream_id: moved_message.stream_id,
-                    topic_name: moved_message.topic,
-                    message_id: moved_message.id,
-                });
             }
 
+            // In case there are no such messages, set it as -1 here, such that it is
+            // neglected in Math.max calcualtion.
+            const latest_locally_available_message_id = event_messages.length
+                ? event_messages.at(-1).id
+                : -1;
+            const latest_locally_unavailable_message_id = messages_not_available_locally.length
+                ? messages_not_available_locally.at(-1)
+                : -1;
+
+            const max_moved_message_id = Math.max(
+                latest_locally_available_message_id,
+                latest_locally_unavailable_message_id,
+            );
+            stream_topic_history.remove_messages({
+                stream_id: old_stream_id,
+                topic_name: orig_topic,
+                max_removed_msg_id: max_moved_message_id,
+                topic_deleted: event.propagate_mode === "change_all",
+            });
+
+            // Add the Recent Conversations entry for the new stream/topics.
+            // We ignore the messages not available locally here as the message_ids
+            // passed in the event may also contain the messages that we might
+            // not have access to like in the case of topic being edited
+            // in a private stream with protected history.
+            const stream_id = stream_changed ? new_stream_id : old_stream_id;
+            const can_latest_message_be_inaccessible =
+                max_moved_message_id === latest_locally_unavailable_message_id &&
+                !sub_store.get(stream_id).history_public_to_subscribers;
```

Comment:
This is basically to handle the cases involving moving multiple messages to a private stream with protected history and some of the messages not being accessible to user due to it. The event includes the IDs of inaccessible messages as well.